### ISSUE TITLE
Code folding not working properly in certain circumstances

### ISSUE THREAD

**Issue by andrewkolos** [Dart-Code/Dart-Code#4121]
**Describe the bug**

At least in top-level functions, code folding can sometimes be unavailable. For example, I might be able to fold a function signature, but not the implementation block (and some but not all blocks within). While this is difficult to describe, it is most easily observed through reproduction or the videos below.

The issue does not appear to happen in Android Studio, so I'm guessing this is not an SDK issue (though VSCode can fold parameter lists, while Android Studio cannot--I'm not sure if this is intended).

**To Reproduce**
1. Clone https://github.com/flutter/flutter (head is f477c8b184d2364fcef0a15cee6ab5c909188229 at the time of writing)
2. Open run_hot.dart (packages/flutter_tools/lib/src)
3. Fold all code. This can be done by opening the command palette (Cmd/Ctrl+Shift+P) and typing _Fold All_.
4. You should see multiple unfolded blocks of code. As a specific example, the body `_defaultReassembleHelper` won't be folded, and `for` loops within might not be folded as well. See screenshots below.

**Expected behavior**
All code blocks should be folded after issuing a _Fold All_ command. Interestingly, copying the contents of the file into a new editor and then using _Fold All_ appears to work as expected. However, saving the file (say, to my Desktop) will unfold the code, and if I try to _Fold All_ again, it is still slightly broken.

**Screenshots**
The issue happening in run_hot.dart

https://user-images.githubusercontent.com/9027551/186756324-d407768f-3190-4b80-92cb-30b4d8e1ce55.mp4

The issue _not_ happening if I copy and paste a problematic function into a new editor:
https://user-images.githubusercontent.com/9027551/186756680-939ba825-7c52-44da-9324-b183e5add931.mp4

The issue happening if I save the new file to disk:
https://user-images.githubusercontent.com/9027551/186756837-88813e9f-1048-409e-9d83-c8e4e06bf0c5.mp4

**Please complete the following information:**
 - Operating System and version: macOS Monterey 12.5.1
 - VS Code version: 1.70.2
 - Dart extension version: 3.46.1
 - Dart/Flutter SDK version: 2.17.6
 - Target device (if the issue relates to Flutter debugging): N/A

I have reproduced this on Windows and another MacBook. Let me know if more information on those environments is needed.

(edit: typo)

**Comment by Hexagonale** [comment]
So the issue is for the functions with multiline parameters. I opened an issue for the VS Code and I think they have provided valuable debug info.

https://github.com/microsoft/vscode/issues/162300

**Comment by DanTup** [comment]
Thanks for the link. I didn't realise this used to work - it looks like there were some changes in VS Code (https://github.com/microsoft/vscode/pull/155302) that stopp it from handling ranges that start/end on the same line.

The data from the server initially contains column numbers (so they don't really intersect at all), but VS Code only supports line folding. I think we'll need to detect this and reduce the first range to end on the line before the next one begins in this case.

**Comment by DanTup** [comment]
Working on a fix at https://dart-review.googlesource.com/c/sdk/+/261900/. I'll close this once it lands. It's a change in the Dart analysis server, so will ship in a future SDK release (rather than a Dart VS Code extension release).

**Comment by methompson** [comment]
I'm still encountering the issues described in the OP. I've installed Flutter / Dart from the beta channel and the pre-release Dart / Flutter VS Code Plugins.

- Dart 2.19.0-255.2.beta
- Flutter 3.4.0-34.1.pre
- Dart-Code v3.53.20221101
- Flutter-Code v3.53.20221101
- VS Code 1.72.2

All running on MacOS 13.0 on a MacBook Pro 16", 2021.

Functions still minimize only on the function definition and NOT the entire function. Please note the screenshots before / after folding.
<img width="245" alt="Screenshot 2022-11-01 at 11 22 38 AM" src="https://user-images.githubusercontent.com/34754528/199284501-a5e6648c-ad48-4c20-b366-5b741dd19bf7.png">
<img width="244" alt="Screenshot 2022-11-01 at 11 22 44 AM" src="https://user-images.githubusercontent.com/34754528/199284519-93951826-6d02-41f0-9559-5f0fd35fb3db.png">

In the above screenshots, `func3` and `func4` should be folded the same as `func1` and `func2`.

I'm uncertain which combination of SDKs, plugins and versions I should be utilizing to get code folding to work correctly.

**Comment by DanTup** [comment]
@methompson looking at the change https://github.com/dart-lang/sdk/commit/604bf2fe3bdfd56868df80b7c9263e9272344e30, it looks like the first Dart SDK it was included in is 2.19.0-257.0.dev:

![Screenshot 2022-11-01 at 18 38 51](https://user-images.githubusercontent.com/1078012/199312123-054598ef-6a8f-4b33-9283-288b898dc8fa.png)

The beta you're on is 2.19.0-255.2.beta so seems like it was cut just before this. Presumably the fix will be in the next beta (I actually thought there had been a more recent beta, but apparently not).

Hope this helps!

**Comment by methompson** [comment]
Thanks for the insight. I've updated to the Master channel, so now I have the following installed on my system, along with the same pre-release versions of dart-code and flutter-code:

- Dart 2.19.0-357.0.dev
- Flutter 3.5.0-12.0.pre.56

I want to note that after installing the stable channel, I completely closed VSCode to make sure any LSPs or daemons would spawn correctly.

Attached is the current state of folding:
<img width="319" alt="Screenshot 2022-11-01 at 2 01 35 PM" src="https://user-images.githubusercontent.com/34754528/199316617-ccdeec29-39cb-4087-9001-3db7b0353027.png">

<img width="288" alt="Screenshot 2022-11-01 at 2 01 42 PM" src="https://user-images.githubusercontent.com/34754528/199316630-4e946af1-5956-415d-ad98-2e600e44f4f5.png">

It seems any code block can be folded, but you cannot fold anything if a foldable code block exists within that code block. As such, there's a definite regression. Note that the class cannot be folded.

func1 can be folded, but as soon as you put a foldable block within a function, you cannot fold the function completely (evidenced by func2). The provided class cannot be folded.

**Comment by DanTup** [comment]
@methompson thanks for the example! There's definitely something wrong there. I've filed #4242 with a cut-down example and will follow up there.

**Comment by TimothyFran** [comment]
Any news on this?

**Comment by DanTup** [comment]
@TimothyFran it's not clear what you're asking about - both this issue and the issue linked in my last comment above were closed several years ago. If something isn't working, please open a new issue with a clear description/screenshot of the problem. Thanks!
